linera-base/src/crypto/secp256k1.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines secp256k1 signature primitives used by the Linera protocol.
+
+use std::fmt;
+
+use secp256k1::{self, Message};
+use serde::{Deserialize, Serialize};
+
+use super::{BcsSignable, CryptoError, CryptoHash, HasTypeName};
+use crate::doc_scalar;
+
+/// A Secp256k1 signature.
+#[derive(Eq, PartialEq, Copy, Clone)]
+pub struct Secp256k1Signature(pub secp256k1::ecdsa::Signature);
+
+impl Secp256k1Signature {
+    /// Computes a secp256k1 signature for `value` using the given `secret`.
+    /// It first serializes the `T` type and then creates the `CryptoHash` from the serialized bytes.
+    pub fn new<'de, T>(value: &T, secret: &secp256k1::SecretKey) -> Self
+    where
+        T: BcsSignable<'de>,
+    {
+        let secp = secp256k1::Secp256k1::new();
+        let message = Message::from_digest(CryptoHash::new(value).as_bytes().0);
+        let signature = secp.sign_ecdsa(&message, secret);
+        Secp256k1Signature(signature)
+    }
+
+    /// Checks a signature.
+    pub fn check<'de, T>(&self, value: &T, author: &secp256k1::PublicKey) -> Result<(), CryptoError>
+    where
+        T: BcsSignable<'de> + fmt::Debug,
+    {
+        let secp = secp256k1::Secp256k1::new();
+        let message = Message::from_digest(CryptoHash::new(value).as_bytes().0);
+        secp.verify_ecdsa(&message, &self.0, author)
+            .map_err(|error| CryptoError::InvalidSignature {
+                error: error.to_string(),
+                type_name: T::type_name().to_string(),
+            })
+    }
+}
+
+impl Serialize for Secp256k1Signature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        if serializer.is_human_readable() {
+            serializer.serialize_str(&hex::encode(self.0.serialize_der()))
+        } else {
+            serializer.serialize_newtype_struct("Signature", &self.0)
```

Comment:
Should we distinguish this from serialized Ed25519 signatures?

Reply:
Most definitely.

Reply:
"Secp256k1Signature"

Reply:
We generally want the serde/rpc formats to match the Rust source code as much as possible.